### PR TITLE
add flag for gpu to call cupy for fast panda

### PR DESCRIPTION
**Pull request by dcolinmorgan**

### CODE REVIEW

File: netZooPy/panda/panda.py [hunk 1]
```diff
@@ -405,30 +414,121 @@ def panda_loop(self, correlation_matrix, motif_matrix, ppi_matrix):
         def t_function(x, y=None):
             '''T function.'''
             if y is None:
-                a_matrix = np.dot(x, x.T)
-                s = np.square(x).sum(axis=1)
-                a_matrix /= np.sqrt(s + s.reshape(-1, 1) - np.abs(a_matrix))
+                a_matrix = xp.dot(x, x.T)
```
Comment (marouenbg) [inline netZooPy/panda/panda.py:417]: same here, the `t_function` is used by both gpuPANDA and cpu PANDA, so we don't want to use `xp.dot` in the cpu version. I'd suggest to write a different `t_function` for gpu or have an `if` statement here

File: netZooPy/panda/panda.py [hunk 2]
```diff
@@ -405,30 +414,121 @@ def panda_loop(self, correlation_matrix, motif_matrix, ppi_matrix):
...
+                s = xp.square(x).sum(axis=1)
+                a_matrix /= xp.sqrt(s + s.reshape(-1, 1) - xp.abs(a_matrix))
+            else:
+                a_matrix = xp.dot(x, y)
+                a_matrix /= xp.sqrt(xp.square(y).sum(axis=0) + xp.square(x).sum(axis=1).reshape(-1, 1) - xp.abs(a_matrix))
+            return a_matrix
+
+        def update_diagonal(diagonal_matrix, num, alpha, step):
+            '''Update diagonal.'''
+            xp.fill_diagonal(diagonal_matrix, xp.nan)
```
Comment (marouenbg) [inline netZooPy/panda/panda.py:427]: same for `update_diagonal`

File: netZooPy/panda/panda.py [hunk 3]
```diff
@@ -405,30 +414,121 @@ def panda_loop(self, correlation_matrix, motif_matrix, ppi_matrix):
...
+            diagonal_std = xp.nanstd(diagonal_matrix, 1)
+            diagonal_fill = diagonal_std * num * math.exp(2 * alpha * step)
+            xp.fill_diagonal(diagonal_matrix, diagonal_fill)
+
+        def gt_function(x, y=None):
```
Comment (marouenbg) [inline netZooPy/panda/panda.py:432]: I think you created `gt_function` as gpu version of `t_function`, in this case we probably should be the `t_function` as in the original version

File: netZooPy/panda/panda.py [hunk 4]
```diff
@@ -405,30 +414,121 @@ def panda_loop(self, correlation_matrix, motif_matrix, ppi_matrix):
...
+                # a_matrix = tf.linalg.matmul(x, y,axes=1)
+                a_matrix = tf.linalg.matmul(x, y)
+                a_matrix /= xp.sqrt(xp.square(y).sum(axis=0) + xp.square(x).sum(axis=1).reshape(-1, 1) - xp.abs(a_matrix.numpy()))
+            return a_matrix.numpy()
+
+        panda_loop_time = time.time()
+        num_tfs, num_genes = motif_matrix.shape
+        step = 0
+        hamming = 1
+        alpha = 0.1
```
Comment (marouenbg) [inline netZooPy/panda/panda.py:449]: I can't believe that alpha is hard-coded in pypanda, we would have been better off writing our own software than fixing this

File: netZooPy/panda/panda.py [hunk 5]
```diff
@@ -405,30 +414,121 @@ def panda_loop(self, correlation_matrix, motif_matrix, ppi_matrix):
...
+            motif = self.motif_matrix_unnormalized.flatten(order='F')
+            force = self.motif_matrix.flatten(order='F')
+            self.export_panda_results = pd.DataFrame({'tf':tfs, 'gene': genes,'motif': motif, 'force': force})
+            #self.export_panda_results = np.column_stack((tfs,genes,motif,force))
+        return motif_matrix
+
+    def GPU_panda_loop(self, correlation_matrix, motif_matrix, ppi_matrix):
+        """Panda algorithm.
+        """
+        def t_function(x, y=None):
```
Comment (marouenbg) [inline netZooPy/panda/panda.py:501]: why is there another `t_function` here I am confused

File: netZooPy/panda/panda.py (hunk 3 above)
Comment (dcolinmorgan) [reply netZooPy/panda/panda.py:432]: this xp is what i was trying to get at-- and explains your last comment on why there is another tfunction -- cupy has the capability to notice IF a gpu is present and then set xp to either np or cp -- making the code agnostic. see the bottom of this page. i tried but couldnt get colab to recognize the gpu automatically via xp (ie xp would = cp) even tho cp commands would work when explicitly made. So I figured I would PR the state it is in (requiring flag) to show you and then tinker with it in future remove the flag once xp = cp.get_array_module(2) sets xp to "cp" and not "np"

File: netZooPy/panda/panda.py [hunk 6]
```diff
@@ -3,21 +3,27 @@
 import math
 import time
 import pandas as pd
-import numpy as np
 from scipy.stats import zscore
 from .timer import Timer
+import numpy as np
+#requires version 7.4.0 and above, via:
+# pip install cupy-cuda101==7.4.0, depending on CUDA version, ie 10.1
+import cupy as cp
```
[the remark this answers is not in the document]
Comment (dcolinmorgan) [reply netZooPy/panda/panda.py:11]: how about just loading cupy in GPU_panda_loop?

[the remark this answers is not in the document]
Comment (dcolinmorgan) [reply netZooPy/panda/panda.py:11]: oh yeah or i could just put if statement where tfunction and update_diag are called

[the remark this answers is not in the document]
Comment (marouenbg) [reply netZooPy/panda/panda.py:11]: yes both are ok, loading cupy in GPU_panda_loop seems cleaner

**Pull request closed without merging**